core/consensus/grandpa/impl/voting_round_impl.cpp

```diff
@@ -0,0 +1,689 @@
+        prevote_timer_{*io_context_},
+        precommit_timer_{*io_context_},
+        logger_{common::createLogger("Grandpa")},
+        prevote_equivocators_(voter_set_->size(), false),
+        precommit_equivocators_(voter_set_->size(), false) {
+    BOOST_ASSERT(voter_set_ != nullptr);
+    BOOST_ASSERT(vote_crypto_provider_ != nullptr);
+    BOOST_ASSERT(prevotes_ != nullptr);
+    BOOST_ASSERT(precommits_ != nullptr);
+    BOOST_ASSERT(env_ != nullptr);
+    BOOST_ASSERT(graph_ != nullptr);
+    BOOST_ASSERT(clock_ != nullptr);
+
+    threshold_ = getThreshold(voter_set_);
+    start_time_ = clock_->now();
+  }
+
+  bool VotingRoundImpl::isPrimary(const Id &id) const {
+    auto index = round_number_ % voter_set_->size();
+    return voter_set_->voters().at(index) == id;
+  }
+
+  bool VotingRoundImpl::isPrimary() const {
+    return isPrimary(id_);
+  }
+
+  size_t VotingRoundImpl::getThreshold(
+      const std::shared_ptr<VoterSet> &voters) {
+    // calculate supermajority
+    auto faulty = (voters->totalWeight() - 1) / 3;
+    return voters->totalWeight() - faulty;
+  }
+
+  void VotingRoundImpl::onFinalize(const Fin &f) {
+    // validate message
+    if (validate(f.vote, f.justification)) {
+      // finalize to state
+      auto finalized = env_->finalize(f.vote.block_hash, f.justification);
+      if (not finalized) {
+        logger_->error(
+            "Could not finalize block {} from round {} with error: {}",
+            f.vote.block_hash.toHex(),
+            f.round_number,
+            finalized.error().message());
+        return;
+      }
+      env_->onCompleted(CompletedRound{.round_number = round_number_,
+                                       .state = cur_round_state_});
+    } else {
+      env_->onCompleted(VotingRoundError::FIN_VALIDATION_FAILED);
+    }
+  }
+
+  bool VotingRoundImpl::validate(
+      const BlockInfo &vote, const GrandpaJustification &justification) const {
+    size_t total_weight = 0;
+    for (const auto &signed_precommit : justification.items) {
+      // verify signatures
+      if (not vote_crypto_provider_->verifyPrecommit(signed_precommit)) {
+        logger_->error(
+            "Received invalid signed precommit during the round {} from the "
+            "peer {}",
+            round_number_,
+            signed_precommit.id.toHex());
+        return false;
+      }
+
+      // check that every signed precommit corresponds to the vote (i.e.
+      // signed_precommits are descendants of the vote). If so add weight of
+      // that voter to the total weight
+      if (env_->getAncestry(vote.block_hash,
+                            signed_precommit.message.block_hash)) {
+        total_weight +=
+            voter_set_->voterWeight(signed_precommit.id)
+                .value_or(0);  // add zero if such voter does not exist
+      }
+    }
+
+    return total_weight >= threshold_;
+  }
+
+  bool VotingRoundImpl::tryFinalize() {
+    if (not completable()) {
+      return false;
```

Comment:
it is normal situation. We try to finalize after each precommit. If there are not enough precommits, round is not completable.
So no warning is needed

Reply:
then add comment pls

Reply:
what kind of comments? we stop function because round is not completable? There is a definintion for completability which explains the thing that I explained in previous comment